rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -5233,6 +5233,102 @@ extern "C" rmw_ret_t rmw_count_subscribers(
   return common_context->graph_cache.get_reader_count(mangled_topic_name, count);
 }
 
+extern "C" rmw_ret_t rmw_count_clients(
+  const rmw_node_t * node, const char * service_name,
+  size_t * count)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_name, RMW_RET_INVALID_ARGUMENT);
+  int validation_result = RMW_TOPIC_VALID;
+  rmw_ret_t ret = rmw_validate_full_topic_name(service_name, &validation_result, nullptr);
+  if (RMW_RET_OK != ret) {
+    return ret;
+  }
+  if (RMW_TOPIC_VALID != validation_result) {
+    const char * reason = rmw_full_topic_name_validation_result_string(validation_result);
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("service_name argument is invalid: %s", reason);
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  RMW_CHECK_ARGUMENT_FOR_NULL(count, RMW_RET_INVALID_ARGUMENT);
+  auto common_context = &node->context->impl->common;
+  const std::string mangled_rq_service_name =
+    make_fqtopic(ROS_SERVICE_REQUESTER_PREFIX, service_name, "Request", false);
+  const std::string mangled_rp_service_name =
+    make_fqtopic(ROS_SERVICE_RESPONSE_PREFIX, service_name, "Reply", false);
+
+  size_t number_of_request_publishers = 0;
+  ret = common_context->graph_cache.get_writer_count(
+    mangled_rq_service_name,
+    &number_of_request_publishers);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  size_t number_of_response_subscribers = 0;
+  ret = common_context->graph_cache.get_reader_count(
+    mangled_rp_service_name,
+    &number_of_response_subscribers);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  if (number_of_request_publishers != number_of_response_subscribers) {
+    return RMW_RET_ERROR;
```

Comment:
I've searched methods to get the size of client ids.

for this, to my knowledge, a type of the`dds_entity_t` should be needed like below.
https://github.com/ros2/rmw_cyclonedds/blob/5b67ae471ba81560b51bac6b1dacfbf271e0dd22/rmw_cyclonedds_cpp/src/rmw_node.cpp#L5110-L5116

however, the function i made have no topic entity and i tried to find it (or the type of `rmw_client_t` ) using `node` and `service_name` provided by the input parameter of the function, but i'd not find it.

to implement the service count, there are two ways in my idea.
1. search all nodes  (https://github.com/ros2/rmw_cyclonedds/blob/5b67ae471ba81560b51bac6b1dacfbf271e0dd22/rmw_cyclonedds_cpp/src/rmw_node.cpp#L4963-L4966)
and count the matched service. 
(https://github.com/ros2/rmw_cyclonedds/blob/5b67ae471ba81560b51bac6b1dacfbf271e0dd22/rmw_cyclonedds_cpp/src/rmw_node.cpp#L5387-L5392) 

2. immediately return `number_of_response_subscribers` without comparing with `number_of_request_publishers` https://github.com/ros2/rmw_cyclonedds/blob/5b67ae471ba81560b51bac6b1dacfbf271e0dd22/rmw_cyclonedds_cpp/src/rmw_node.cpp#L5233 

What should I do? could you tell me if there's a good way I haven't found?

Reply:
It really depends on what you want to use it for, right?

If all it is ever going to be used for is for a user to get an indication of the number of clients/services from the CLI, then I'd say just look at the number of response subscribers and be done with it. (It would also make it really trivial to have a single implementation for counting clients and subscribers: all you need to do is pass in the topic name pattern you're looking for.)

If you do want to protect against all kinds of things that might happen if people start to write parts of the system in DDS without using the ROS libraries, I'd go a step further. If you want to have a well-defined notion of what number you actually return (there is no such thing as the "current" number of subscribers in a distributed system that's not in steady state) then the first problem is to define what exactly the returned value means.

I suspect it is, at least for now, the first case. In that case, it is probably best to keep it as simple as possible.

I just checked the related PRs to get a better sense of the goal, and I noticed that the comment in the RMW says "does not allocate memory". I'd say that's not true, e.g., just constructing the topic node likely involves memory allocation.

Reply:
thank you for your information and sorry for misunderstanding 
> I suspect it is, at least for now, the first case. In that case, it is probably best to keep it as simple as possible.

Does **the first case** mean just looking at the number of response subscribers ?

Reply:
i just saw comments again. i don't yet have a deep understanding of DDS and ROS. 
my current estimate is below.

>2. immediately return number_of_response_subscribers  without comparing with number_of_request_publishers.
> https://github.com/ros2/rmw_cyclonedds/blob/5b67ae471ba81560b51bac6b1dacfbf271e0dd22/rmw_cyclonedds_cpp/src/rmw_node.cpp#L5233 

this just get the number of data readers for a DDS topic. it can't protect for those to use reading/writing parts in DDS without ROS as you mentioned.



>1. https://github.com/ros2/rmw_cyclonedds/blob/5b67ae471ba81560b51bac6b1dacfbf271e0dd22/rmw_cyclonedds_cpp/src/rmw_node.cpp#L5387-L5392

in this case, it can get the numbers by Node which is a concept of ROS, so this is likely to protect from reading/writing parts in DDS without using the ROS



> I suspect it is, at least for now, the first case. In that case, it is probably best to keep it as simple as possible.

this means the way which get the numbers using Node, right? 

i'll post PR again as soon as you respond.

Reply:
@eboasson Friendly ping, I think we are waiting for a response here.

Reply:
Indeed ... sorry that I let it slip through the cracks ...

I think for the purpose of showing the number of clients and services in the `ros2` tool, there is no harm in assuming that every part of the system is either built on ROS 2, or plays by the rules that ROS 2 follows. That is, that there are no other readers/writers for the service topics.

I also think it is fine to not worry about all the edge cases one I come up with (years of practice there ...), again, given the purpose. I would put a note in the documentation that the number is only reliable "in steady state", or something to that effect, just so people will realise there are limitations to these operations.

And then, just go with: `return common_context->graph_cache.get_reader_count(mangled_topic_name, count);`

Does that sound reasonable?

Reply:
absolutely.
I fixed.

Reply:
ping @eboasson, could you review this again? 
I'm going to quickly update other PRs related to this upon your acceptance.